no-extra-parens conflicts with wrap-iife

<!--
    ESLint adheres to the [JS Foundation Code of Conduct](https://js.foundation/community/code-of-conduct).

    This template is for bug reports. If you are here for another reason, please see below:

    1. To propose a new rule: https://eslint.org/docs/developer-guide/contributing/new-rules
    2. To request a rule change: https://eslint.org/docs/developer-guide/contributing/rule-changes
    3. To request a change that is not a bug fix, rule change, or new rule: https://eslint.org/docs/developer-guide/contributing/changes
    4. If you have any questions, please stop by our chatroom: https://gitter.im/eslint/eslint

    Note that leaving sections blank will make it difficult for us to troubleshoot and we may have to close the issue.
-->

**Tell us about your environment**

<!--
    If you are using ESLint v6.5.0 or later, you can run ESLint with the `--env-info` flag and paste the output here.
-->

* **ESLint Version:** 7.0.0-alpha.0
* **Node Version:** v12.14.0
* **npm Version:** v6.13.4

**What parser (default, Babel-ESLint, etc.) are you using?**

default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

<!-- Paste your configuration below: -->
```js
module.exports = {
    parserOptions: {
        ecmaVersion: 2015
    },
};
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

<!-- Paste the source code below: -->
```js
/* eslint wrap-iife: ["error", "inside", { functionPrototypeMethods: true }] */
/* eslint no-extra-parens: ["error"] */

var x = function(){}.call(); // wrap-iife error

var x = (function(){}).call(); // no-extra-parens error
```

[Online Demo Link](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IHdyYXAtaWlmZTogW1wiZXJyb3JcIiwgXCJpbnNpZGVcIiwgeyBmdW5jdGlvblByb3RvdHlwZU1ldGhvZHM6IHRydWUgfV0gKi9cbi8qIGVzbGludCBuby1leHRyYS1wYXJlbnM6IFtcImVycm9yXCJdICovXG5cbnZhciB4ID0gZnVuY3Rpb24oKXt9LmNhbGwoKTsgLy8gd3JhcC1paWZlIGVycm9yXG5cbnZhciB4ID0gKGZ1bmN0aW9uKCl7fSkuY2FsbCgpOyAvLyBuby1leHRyYS1wYXJlbnMgZXJyb3IiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjYsInNvdXJjZVR5cGUiOiJzY3JpcHQiLCJlY21hRmVhdHVyZXMiOnt9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=) (demo is v6.8.0, but it's same)

<!-- Paste the command you used to run ESLint: -->
```bash
eslint index.js --fix
```

**What did you expect to happen?**

These are two same lines, except for the parens around the function expression.

`wrap-iife` requires these parens, while `no-extra-parens` disallows them.

I'm not sure what should be expected default behavior for `no-extra-parens`:

> This rule always ignores extra parentheses around the following:
>
> immediately-invoked function expressions (also known as IIFEs) such as var x = (function () {})(); and ((function foo() {return 1;})()) to avoid conflicts with the wrap-iife rule

This isn't quite an IIFE, but it is a conflict with `wrap-iife`.

**What actually happened? Please include the actual, raw output from ESLint.**

Autofix worked 10 times, so the code at the end remained unchanged, with the following errors:

```
  4:9  error  Wrap an immediate function invocation in parentheses  wrap-iife
  6:9  error  Unnecessary parentheses around expression             no-extra-parens
```

If it was set to run an odd number of times, it would basically flip those two lines.

**Are you willing to submit a pull request to fix this bug?**

Yes. I'm not sure should this be default or an option.

--- a/docs/rules/no-extra-parens.md
+++ b/docs/rules/no-extra-parens.md
@@ -7,7 +7,7 @@ This rule restricts the use of parentheses to only where they are necessary.
 This rule always ignores extra parentheses around the following:
 
 * RegExp literals such as `(/abc/).test(var)` to avoid conflicts with the [wrap-regex](wrap-regex.md) rule
-* immediately-invoked function expressions (also known as IIFEs) such as `var x = (function () {})();` and `((function foo() {return 1;})())` to avoid conflicts with the [wrap-iife](wrap-iife.md) rule
+* immediately-invoked function expressions (also known as IIFEs) such as `var x = (function () {})();` and `var x = (function () {}());` to avoid conflicts with the [wrap-iife](wrap-iife.md) rule
 * arrow function arguments to avoid conflicts with the [arrow-parens](arrow-parens.md) rule
 
 ## Options
@@ -26,6 +26,7 @@ This rule has an object option for exceptions to the `"all"` option:
 * `"enforceForArrowConditionals": false` allows extra parentheses around ternary expressions which are the body of an arrow function
 * `"enforceForSequenceExpressions": false` allows extra parentheses around sequence expressions
 * `"enforceForNewInMemberExpressions": false` allows extra parentheses around `new` expressions in member expressions
+* `"enforceForFunctionPrototypeMethods": false` allows extra parentheses around immediate `.call` and `.apply` method calls on function expressions and around function expressions in the same context.
 
 ### all
 
@@ -220,6 +221,22 @@ const foo = (new Bar()).baz;
 const quux = (new Bar())[baz];
 
 (new Bar()).doSomething();
+```
+
+### enforceForFunctionPrototypeMethods
+
+Examples of **correct** code for this rule with the `"all"` and `{ "enforceForFunctionPrototypeMethods": false }` options:
+
+```js
+/* eslint no-extra-parens: ["error", "all", { "enforceForFunctionPrototypeMethods": false }] */
+
+const foo = (function () {}).call();
+
+const bar = (function () {}).apply();
+
+const baz = (function () {}.call());
+
+const quux = (function () {}.apply());
 ```
 
 ### functions

--- a/lib/rules/no-extra-parens.js
+++ b/lib/rules/no-extra-parens.js
@@ -51,7 +51,8 @@ module.exports = {
                                 ignoreJSX: { enum: ["none", "all", "single-line", "multi-line"] },
                                 enforceForArrowConditionals: { type: "boolean" },
                                 enforceForSequenceExpressions: { type: "boolean" },
-                                enforceForNewInMemberExpressions: { type: "boolean" }
+                                enforceForNewInMemberExpressions: { type: "boolean" },
+                                enforceForFunctionPrototypeMethods: { type: "boolean" }
                             },
                             additionalProperties: false
                         }
@@ -83,11 +84,27 @@ module.exports = {
             context.options[1].enforceForSequenceExpressions === false;
         const IGNORE_NEW_IN_MEMBER_EXPR = ALL_NODES && context.options[1] &&
             context.options[1].enforceForNewInMemberExpressions === false;
+        const IGNORE_FUNCTION_PROTOTYPE_METHODS = ALL_NODES && context.options[1] &&
+            context.options[1].enforceForFunctionPrototypeMethods === false;
 
         const PRECEDENCE_OF_ASSIGNMENT_EXPR = precedence({ type: "AssignmentExpression" });
         const PRECEDENCE_OF_UPDATE_EXPR = precedence({ type: "UpdateExpression" });
 
         let reportsBuffer;
+
+        /**
+         * Determines whether the given node is a `call` or `apply` method call, invoked directly on a `FunctionExpression` node.
+         * Example: function(){}.call()
+         * @param {ASTNode} node The node to be checked.
+         * @returns {boolean} True if the node is an immediate `call` or `apply` method call.
+         * @private
+         */
+        function isImmediateFunctionPrototypeMethodCall(node) {
+            return node.type === "CallExpression" &&
+                node.callee.type === "MemberExpression" &&
+                node.callee.object.type === "FunctionExpression" &&
+                ["call", "apply"].includes(astUtils.getStaticPropertyName(node.callee));
+        }
 
         /**
          * Determines if this rule should be enforced for a node given the current configuration.
@@ -122,6 +139,10 @@ module.exports = {
             }
 
             if (node.type === "SequenceExpression" && IGNORE_SEQUENCE_EXPRESSIONS) {
+                return false;
+            }
+
+            if (isImmediateFunctionPrototypeMethodCall(node) && IGNORE_FUNCTION_PROTOTYPE_METHODS) {
                 return false;
             }
 
@@ -924,7 +945,12 @@ module.exports = {
             LogicalExpression: checkBinaryLogical,
 
             MemberExpression(node) {
-                const nodeObjHasExcessParens = hasExcessParens(node.object);
+                const nodeObjHasExcessParens = hasExcessParens(node.object) &&
+                    !(
+                        isImmediateFunctionPrototypeMethodCall(node.parent) &&
+                        node.parent.callee === node &&
+                        IGNORE_FUNCTION_PROTOTYPE_METHODS
+                    );
 
                 if (
                     nodeObjHasExcessParens &&

--- a/tests/lib/rules/no-extra-parens.js
+++ b/tests/lib/rules/no-extra-parens.js
@@ -477,6 +477,27 @@ ruleTester.run("no-extra-parens", rule, {
         { code: "(new foo.bar()).baz()", options: ["all", { enforceForNewInMemberExpressions: false }] },
         { code: "((new foo.bar())).baz()", options: ["all", { enforceForNewInMemberExpressions: false }] },
 
+        // ["all", { enforceForFunctionPrototypeMethods: false }]
+        { code: "var foo = (function(){}).call()", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+        { code: "var foo = (function(){}).apply()", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+        { code: "var foo = (function(){}.call())", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+        { code: "var foo = (function(){}.apply())", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+        { code: "var foo = (function(){}).call(arg)", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+        { code: "var foo = (function(){}.apply(arg))", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+        { code: "var foo = (function(){}['call']())", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+        { code: "var foo = (function(){})[`apply`]()", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+        { code: "var foo = ((function(){})).call()", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+        { code: "var foo = ((function(){}).apply())", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+        { code: "var foo = ((function(){}.call()))", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+        { code: "var foo = ((((function(){})).apply()))", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+        { code: "foo((function(){}).call().bar)", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+        { code: "foo = (function(){}).call()()", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+        { code: "foo = (function(){}.call())()", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+        { code: "var foo = { bar: (function(){}.call()) }", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+        { code: "var foo = { [(function(){}.call())]: bar  }", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+        { code: "if((function(){}).call()){}", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+        { code: "while((function(){}.apply())){}", options: ["all", { enforceForFunctionPrototypeMethods: false }] },
+
         "let a = [ ...b ]",
         "let a = { ...b }",
         {
@@ -1256,6 +1277,316 @@ ruleTester.run("no-extra-parens", rule, {
                 {
                     messageId: "unexpected",
                     type: "NewExpression"
+                }
+            ]
+        },
+
+        // enforceForFunctionPrototypeMethods
+        {
+            code: "var foo = (function(){}).call()",
+            output: "var foo = function(){}.call()",
+            options: ["all"],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "FunctionExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (function(){}.apply())",
+            output: "var foo = function(){}.apply()",
+            options: ["all"],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "CallExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (function(){}).apply()",
+            output: "var foo = function(){}.apply()",
+            options: ["all", {}],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "FunctionExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (function(){}.call())",
+            output: "var foo = function(){}.call()",
+            options: ["all", {}],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "CallExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (function(){}).call()",
+            output: "var foo = function(){}.call()",
+            options: ["all", { enforceForFunctionPrototypeMethods: true }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "FunctionExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (function(){}).apply()",
+            output: "var foo = function(){}.apply()",
+            options: ["all", { enforceForFunctionPrototypeMethods: true }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "FunctionExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (function(){}.call())",
+            output: "var foo = function(){}.call()",
+            options: ["all", { enforceForFunctionPrototypeMethods: true }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "CallExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (function(){}.apply())",
+            output: "var foo = function(){}.apply()",
+            options: ["all", { enforceForFunctionPrototypeMethods: true }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "CallExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (function(){}.call)()", // removing these parens does not cause any conflicts with wrap-iife
+            output: "var foo = function(){}.call()",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "MemberExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (function(){}.apply)()", // removing these parens does not cause any conflicts with wrap-iife
+            output: "var foo = function(){}.apply()",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "MemberExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (function(){}).call",
+            output: "var foo = function(){}.call",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "FunctionExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (function(){}.call)",
+            output: "var foo = function(){}.call",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "MemberExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = new (function(){}).call()",
+            output: "var foo = new function(){}.call()",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "FunctionExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (new function(){}.call())",
+            output: "var foo = new function(){}.call()",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "NewExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (function(){})[call]()",
+            output: "var foo = function(){}[call]()",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "FunctionExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (function(){}[apply]())",
+            output: "var foo = function(){}[apply]()",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "CallExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (function(){}).bar()",
+            output: "var foo = function(){}.bar()",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "FunctionExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (function(){}.bar())",
+            output: "var foo = function(){}.bar()",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "CallExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (function(){}).call.call()",
+            output: "var foo = function(){}.call.call()",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "FunctionExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (function(){}.call.call())",
+            output: "var foo = function(){}.call.call()",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "CallExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (call())",
+            output: "var foo = call()",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "CallExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (apply())",
+            output: "var foo = apply()",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "CallExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = (bar).call()",
+            output: "var foo = bar.call()",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "Identifier"
+                }
+            ]
+        },
+        {
+            code: "var foo = (bar.call())",
+            output: "var foo = bar.call()",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "CallExpression"
+                }
+            ]
+        },
+        {
+            code: "((() => {}).call())",
+            output: "(() => {}).call()",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "CallExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = function(){}.call((a.b))",
+            output: "var foo = function(){}.call(a.b)",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "MemberExpression"
+                }
+            ]
+        },
+        {
+            code: "var foo = function(){}.call((a).b)",
+            output: "var foo = function(){}.call(a.b)",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "Identifier"
+                }
+            ]
+        },
+        {
+            code: "var foo = function(){}[('call')]()",
+            output: "var foo = function(){}['call']()",
+            options: ["all", { enforceForFunctionPrototypeMethods: false }],
+            errors: [
+                {
+                    messageId: "unexpected",
+                    type: "Literal"
                 }
             ]
         },

